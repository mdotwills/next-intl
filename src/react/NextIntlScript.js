/* global __NEXT_INTL_POLYFILLS__:false */
import React from 'react';

// The script bellow was minified using terser: https://xem.github.io/terser-online/
// You may find the original script below in case you need to modify it:
/*
(function (polyfill) {
    const shouldPolyfill = new Function('var exports={};' + polyfill.shouldPolyfill + 'return shouldPolyfill()');

    if (shouldPolyfill()) {
        var currentScript = document.currentScript;
        var polyfillScript = document.createElement('script');

        polyfillScript.src = polyfill.asset;

        currentScript.parentNode.insertBefore(polyfillScript, currentScript.nextSibling);
    }
})(polyfill);
*/
const renderScript = (polyfill) => `!function(e){if(new Function("var exports={};"+e.shouldPolyfill+"return shouldPolyfill()")()){var n=document.currentScript,t=document.createElement("script");t.src=e.asset,n.parentNode.insertBefore(t,n.nextSibling)}}(${JSON.stringify(polyfill)})`;

const NextIntlScript = () => {
    if (typeof __NEXT_INTL_POLYFILLS__ === 'undefined') {
        throw new Error('Could not locale polyfills data, did you forgot to enable the plugin in the next.config.js file?');
    }

    return (
        <>
            { __NEXT_INTL_POLYFILLS__.map((polyfill) => (
                <script
                    key={ polyfill.asset }
                    dangerouslySetInnerHTML={ { __html: renderScript(polyfill) } } />
            )) }
        </>
    );
};

export default NextIntlScript;
