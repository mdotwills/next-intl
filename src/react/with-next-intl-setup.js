import React, { Component } from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import App from 'next/app';
import NextIntlProvider, { getInitialProps, toInitialProps } from './NextIntlProvider';

const withNextIntlSetup = (config) => (WrappedApp) => {
    let locale;
    let nextIntlProviderProps;

    class WithNextIntlSetup extends Component {
        static propTypes = {
            appProps: PropTypes.object,
            nextIntlProviderProps: PropTypes.object,
        };

        static displayName = `withNextIntlSetup(${WrappedApp.displayName || WrappedApp.name || /* istanbul ignore next */ 'App'})`;

        static async getInitialProps(appCtx) {
            if (typeof window === 'undefined') {
                nextIntlProviderProps = await getInitialProps(config, appCtx.ctx);
                locale = config.locales.find(({ id }) => id === nextIntlProviderProps.initialData.localeId);
            }

            appCtx.ctx.locale = locale;

            const appProps = await (WrappedApp.getInitialProps ? WrappedApp.getInitialProps(appCtx) : App.getInitialProps(appCtx));

            return { appProps, nextIntlProviderProps };
        }

        render() {
            const { appProps, nextIntlProviderProps, ...rest } = this.props;

            return (
                <NextIntlProvider
                    ref={ this.providerRef }
                    { ...config }
                    { ...nextIntlProviderProps }
                    onChange={ this.handleChange }>
                    <WrappedApp { ...appProps } { ...rest } />
                </NextIntlProvider>
            );
        }

        providerRef = (provider) => {
            if (provider) {
                nextIntlProviderProps = toInitialProps(provider);
                locale = provider.getValue().locale;
            } else {
                nextIntlProviderProps = locale = undefined;
            }
        };

        handleChange = ({ locale: newLocale }) => {
            locale = newLocale;
        };
    }

    hoistNonReactStatics(WithNextIntlSetup, WrappedApp, { getInitialProps: true });

    return WithNextIntlSetup;
};

export default withNextIntlSetup;
