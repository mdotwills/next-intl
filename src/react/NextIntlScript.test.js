import React from 'react';
import { render } from '@testing-library/react';
import NextIntlScript from './NextIntlScript';

beforeEach(() => {
    global.__NEXT_INTL_POLYFILLS__ = [
        {
            asset: 'intl-polyfill-1.js',
            shouldPolyfill: 'exports.shouldPolyfill = () => true',
        },
        {
            asset: 'intl-polyfill-2.js',
            shouldPolyfill: 'exports.shouldPolyfill = () => false',
        },
    ];
});

afterEach(() => {
    console.error.mockRestore?.();
});

it('should render a script tag with conditional loading for each polyfill correctly', () => {
    const { container } = render(
        <NextIntlScript />,
    );

    const script = container.querySelector('script');

    expect(script.innerHTML).toMatchSnapshot();
});

it('should throw if unable to find polyfill chunk', () => {
    delete global.__NEXT_INTL_POLYFILLS__;

    jest.spyOn(console, 'error').mockImplementation(() => {});

    expect(() => {
        render(
            <NextIntlScript />,
        );
    }).toThrow(new Error('Could not locale polyfills data, did you forgot to enable the plugin in the next.config.js file?'));
});
