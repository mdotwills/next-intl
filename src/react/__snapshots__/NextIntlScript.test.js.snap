// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a script tag with conditional loading for each polyfill correctly 1`] = `"!function(e){if(new Function(\\"var exports={};\\"+e.shouldPolyfill+\\"return shouldPolyfill()\\")()){var n=document.currentScript,t=document.createElement(\\"script\\");t.src=e.asset,n.parentNode.insertBefore(t,n.nextSibling)}}({\\"asset\\":\\"intl-polyfill-1.js\\",\\"shouldPolyfill\\":\\"exports.shouldPolyfill = () => true\\"})"`;
